File: funções/derivada.js
function derivada(funcao) {
    return funcao.map((termo) => {
        if (termo.tipo === "polinomial") {
            return termo.exp === 0 
            ? { tipo: "polinomial", coef: 0, exp: 0 }
            : { tipo: "polinomial", coef: termo.coef * termo.exp, exp: termo.exp - 1 };
        } else if (termo.tipo === "exponencial") {
            return {
                tipo: "exponencial",
                coef: termo.coef * Math.log(termo.base),
                base: termo.base
            }
        }
    });
}

module.exports = derivada;

File: py/funcoes/derivada.py
import math

def derivada(funcao):
    derivada_funcao = []
    for termo in funcao:
        if termo["tipo"] == "polinomial":
            if termo["expoente"] == 0:
                derivada_funcao.append({
                    "tipo": "polinomial",
                    "coeficiente": 0,
                    "expoente": 0
                })
            else:
                derivada_funcao.append({
                    "tipo": "polinomial",
                    "coeficiente": termo["coeficiente"] * termo["expoente"],
                    "expoente": termo["expoente"] - 1
                })
        elif termo["tipo"] == "exponencial":
            derivada_funcao.append({
                "tipo": "exponencial",
                "coeficiente": termo["coeficiente"] * math.log(termo["base"]),
                "base": termo["base"]
            })
    return derivada_funcao